chalk-solve/src/clauses/builtin_traits/fn_family.rs

```diff
@@ -0,0 +1,87 @@
+use crate::clauses::ClauseBuilder;
+use crate::infer::instantiate::IntoBindersAndValue;
+use crate::rust_ir::WellKnownTrait;
+use crate::{Interner, RustIrDatabase, TraitRef};
+use chalk_ir::{
+    AliasTy, ApplicationTy, Binders, Normalize, ProjectionTy, Substitution, TraitId, Ty, TyData,
+    TypeName, VariableKinds,
+};
+
+/// Handles clauses for FnOnce/FnMut/Fn.
+/// If `self_ty` is a function, we push a clause of the form
+/// `fn(A1, A2, ..., AN) -> O: FnTrait<(A1, A2, ..., AN)>`, where `FnTrait`
+/// is the trait corresponding to `trait_id` (FnOnce/FnMut/Fn)
+///
+/// If `trait_id` is `FnOnce`, we also push a clause for the output type of the form:
+/// `Normalize(<fn(A) -> B as FnOnce<(A,)>>::Output -> B)`
+/// We do not add the usual `Implemented(fn(A) -> b as FnOnce<(A,)>` clause
+/// as a condition, since we already called `push_fact` with it
+pub fn add_fn_trait_program_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    trait_id: TraitId<I>,
+    self_ty: Ty<I>,
+) {
+    let interner = db.interner();
+    match self_ty.data(interner) {
+        TyData::Function(fn_val) => {
```

Comment:
Hmm, so long as the `Fn*` traits are 'non-enumerable', this is correct, but otherwise we should also push these clauses if self-ty is an inference variable or an alias (aliases should be fixed by the work on semantics vs syntactic equality), since it's possible they could apply.

Reply:
That's a general problem with the built-in impls currently, I think

Reply:
As I noted below, I think the right answer is floundering -- this is indeed why we added floundering in the first place.